### PR TITLE
[None][fix] Canonicalize small-M BF16 FC1 execution

### PR DESCRIPTION
**Pull request by esha-nvidia**
@coderabbitai summary

## Description

Speculative verification evaluates `K + 1` rows in one target forward. For
small draft widths, this gives the unquantized BF16 gate/up projection an
`M` dimension between 9 and 15, where cuBLAS can use a different reduction
tactic from target-only decoding. The resulting row-wise BF16 difference can
be large enough to change deterministic greedy logits.

This change adds an opt-in FC1 canonicalization path to `GatedMLP`. When
enabled, an unquantized, two-dimensional BF16 input with 9-15 rows is padded
to 16 rows for the gate/up projection and trimmed immediately afterward. The
DeepSeek V3 family enables the behavior only for the experimentally validated
first dense layer through
`TLLM_DSV3_LAYER0_CANONICAL_SMALL_M_FC1_ROWS=16`.

Compiled execution uses `torch.cond`, avoiding separate graphs at the 8/9/16
row boundaries without cloning inputs on the unpadded paths.

The default remains disabled. Quantized projections, non-BF16 inputs, and
other row counts retain their existing behavior.

## Test Coverage

- Focused module and DeepSeek configuration tests
  - 23 passed in a TensorRT-LLM runtime environment
  - Includes decoder-constructor wiring checks for layer 0 and layer 1
- Dynamic `torch.compile(fullgraph=True)` test verifies one graph across
  M=8,9,12,15,16,17
- Dynamic Inductor compilation passed across the same boundary values
- `python -m py_compile` for both modified modules and the new test
- Full repository pre-commit suite passed on all four changed files
- `git diff --check origin/main...HEAD`
- Runtime validation: the enabled path reproduced the deterministic
  target-only 256-token response under speculative verification; a matched
  16-request panel retained exact target-equivalent outputs.

## PR Checklist

- [x] The description explains the issue and the narrowly scoped solution.
- [x] The change follows the TensorRT-LLM coding guidelines.
- [x] Focused tests cover padded and bypassed paths.
- [x] No new dependencies, ownership changes, or architecture updates are
  required.

## GitHub Bot Help

To see a list of available CI bot commands, comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/modules/gated_mlp.py [hunk 1]
```diff
@@ -260,6 +281,33 @@ def _fused_gate_up_swiglu(self, x, fp4_out=False):
     # because the CTA tile height exceeds the output allocation.
     _FP4OUT_MIN_M = 128
 
+    def _run_gate_up_proj(
+            self, x: Union[torch.Tensor, Fp4QuantizedTensor]) -> torch.Tensor:
+        if (self.small_m_fc1_pad_rows and isinstance(x, torch.Tensor)
+                and x.ndim == 2 and x.dtype == torch.bfloat16
+                and not self.gate_up_proj.has_any_quant):
+            num_rows = x.shape[0]
```
[the remark this answers is not in the document]
Comment (esha-nvidia) [reply tensorrt_llm/_torch/modules/gated_mlp.py:289]: Validated against current head `87d0e07477`. The layer-scope concern was valid on the reviewed `0af07e5b` revision but is stale on the current implementation: `TLLM_DSV3_LAYER0_CANONICAL_SMALL_M_FC1_ROWS` is passed through `_get_layer0_canonical_small_m_fc1_rows(layer_idx)`, which returns `16` only for `layer_idx == 0` and `0` otherwise. I also accepted the review summary’s separate coverage point and added a constructor-level test that exercises env → `DeepseekV3DecoderLayer` → `GatedMLP`, proving layer 0 receives `16` and layer 1 receives `0`. The focused suite is 23/23 passing, and the full repository pre-commit suite passes on all changed files.

**Review by mikeiovine — COMMENTED** [review]
The change makes sense to me but can I know a bit more about the motivation? Just worried that if we go down the path of trying to make spec dec execution bit exact with non spec dec, we'll be fighting a never ending uphill battle due to kernel tactic selection

**Comment by esha-nvidia** [comment]
> The change makes sense to me but can I know a bit more about the motivation? Just worried that if we go down the path of trying to make spec dec execution bit exact with non spec dec, we'll be fighting a never ending uphill battle due to kernel tactic selection

Closing this -- realized during a separate investigation that we were skipping the PyExecutor warmup leading to some bad results, rather than needing this. Apologies for the noise!
